backend/downsample.py

```diff
@@ -0,0 +1,255 @@
+        next_average = [0, 0]
+        for record in buckets[next_index]:
+            next_average[0] += record[0]
+            next_average[1] += record[1]
+        next_average[0] /= len(buckets[next_index])
+        next_average[1] /= len(buckets[next_index])
+
+        # Select record of highest triangle area in each bucket
+        result.append(
+            max(bucket, key=lambda record, next_average=next_average:
+                triangle_area(result[-1], record, next_average)))
+    return result
+
+
+def max_min_downsample(records, method, max_records):
+    """Maximum and Minimum Downsampling strategy
+
+        Downsample the records with max or min strategy specified by parameter.
+
+        Args:
+            records: Records in 1 second
+            method: one of {max, min} to specify the strategy you want to use.
+            max_records: limit of returned records
+
+        Returns:
+            list: downsampled records from the given records with number of
+            NUMBER_OF_RECORDS_PER_SECOND.
+    """
+    if len(records) <= max_records:
+        return records
+
+    timespan = len(records) // max_records
+    result = list()
+    for index in range(max_records):
+        records_in_timespan = records[index * timespan: (index+1)*timespan]
+        if method == 'max':
+            result.append(
+                max(records_in_timespan, key=lambda record: record[1]))
+        elif method == 'min':
+            result.append(
+                min(records_in_timespan, key=lambda record: record[1]))
+    return result
+
+
+def average_downsample(records, max_records):
+    """Average Downsampling strategy
+
+    Downsample the records with average strategy.
+
+    Args:
+        records: records in 1 second
+        max_records: limit of returned records
+
+    Returns:
+        list: downsampled records
+    """
+    if len(records) <= max_records:
+        return records
+
+    timespan = len(records) // max_records
+    result = list()
+    for index in range(max_records):
+        records_in_timespan = records[index * timespan: (index+1)*timespan]
+        average = [0, 0, records_in_timespan[0][2]]
+        for record in records_in_timespan:
+            average[0] += record[0]
+            average[1] += record[1]
+
+        average[0] /= len(records_in_timespan)
+        average[1] /= len(records_in_timespan)
+
+        average[0] = int(average[0])
+        average[1] = round(average[1], FLOAT_PRECISION)
+        result.append(average)
+    return result
+
+
+def downsample(filename, strategy, max_records_per_second):
+    """Read the file and downsample with the given strategy.
+
+        Assume the records file is on local disk, read the records and
+        downsample the records in one-second-at-a-time fashion.
+
+        Args:
+            filename: The name of the records file
+            strategy: Downsampling strategy.
+            max_records_per_second: Number of records to save per second
```

Comment:
How about use decimation_factor instead. 
e.g. With original sampling as 1M Hz and decimation factor as 100000, the output becomes 10 HZ.

Reply:
I don't think it's a good idea. For multiple channels case, if we keep the same factor, there will be no data remaining.
But later as we get to multiple level downsampling, this seems to be a good consideration.

Reply:
Using a constant limit ensures a good visualization and keep the max possible precision if data is not too much in size.